django/core/paginator.py

```diff
@@ -151,7 +151,7 @@ def __len__(self):
 
     def __getitem__(self, index):
         if not isinstance(index, (int, slice)):
-            raise TypeError
+            raise TypeError('Page indices must be integers or slices, not %s.' % type(index).__name__)
```

Comment:
Maybe `index.__class__.__name__`.

Reply:
Why or is this just personal preference?

`type()` is documented as: 

https://docs.python.org/3/library/functions.html#type

> With one argument, return the type of an object. The return value is a type object and generally the same object as returned by `object.__class__`.

Just as we normally use `len()` instead of calling `__len__` directly or we use `str()` instead of calling `__str__` directly, I prefer to use the higher level builtin function unless there is a reason not to. I don't think there is such a reason in this case, but please let me know if I'm wrong.

Reply:
Nothing specific, it's just a pattern commonly used in Django. Probably because it was not the same in Python2.

Reply:
So I'm fine to leave `type()` calls.

Reply:
As a separate PR, I'd be happy to modernize the pattern across the codebase if you'd like. But if you don't feel the need, that is fine too.

Reply:
Did you check `%timeit`? It suggests that `index.__class__.__name__` is faster.

Reply:
@jdufresne I don't think that it's worth to change existing code.

Reply:
> It suggests that index.__class__.__name__ is faster.

I think you would find the same with `len()` vs `__len__` and `str()` vs `__str__`. The overhead is the additional function call. But, under normal circumstances, I don't think we should use that alone as a reason to avoid these high level builtin functions, unless every little bit counts.

In this specific case, this is formatting a string for an error that occurs during incorrect indexing. Therefore, a programming error. So I don't expect this will create any real significant overhead in a practical sense.